### PR TITLE
fix: rewrite Dockerfile to resolve architecture mismatch issues

### LINKED ISSUE #130: Unable to fork /bin/clang.
**Describe the bug**
While I am setting up retina in K8s infra. I am facing below error. 

```ts=2024-03-23T16:55:10.369Z level=debug caller=loader/compile.go:26 msg="Error running command" goversion=go1.21.8 os=linux arch=arm64 numcores=8 hostname=ip-10-149-82-88.ec2.internal podname=retina-agent-4hldd version=v0.0.1 apiserver=https://172.20.0.1:443 plugins=packetforward command="/bin/clang -target bpf -Wall -D__TARGET_ARCH_arm64 -g -O2 -c /go/src/github.com/microsoft/retina/pkg/plugin/packetforward/_cprog/packetforward.c -o /go/src/github.com/microsoft/retina/pkg/plugin/packetforward/packetforward_bpf.o -I/go/src/github.com/microsoft/retina/pkg/plugin/packetforward/../lib/_arm64 -I/go/src/github.com/microsoft/retina/pkg/plugin/packetforward/../lib/common/libbpf/_src" stderr= error="fork/exec /bin/clang: no such file or directory"
ts=2024-03-23T16:55:10.369Z level=info caller=server/server.go:79 msg="gracefully shutting down HTTP server..." goversion=go1.21.8 os=linux arch=arm64 numcores=8 hostname=ip-10-149-82-88.ec2.internal podname=retina-agent-4hldd version=v0.0.1 apiserver=https://172.20.0.1:443 plugins=packetforward
ts=2024-03-23T16:55:10.369Z level=info caller=server/server.go:71 msg="HTTP server stopped with err: http: Server closed" goversion=go1.21.8 os=linux arch=arm64 numcores=8 hostname=ip-10-149-82-88.ec2.internal podname=retina-agent-4hldd version=v0.0.1 apiserver=https://172.20.0.1:443 plugins=packetforward
ts=2024-03-23T16:55:10.369Z level=panic caller=controllermanager/controllermanager.go:118 msg="Error running controller manager" goversion=go1.21.8 os=linux arch=arm64 numcores=8 hostname=ip-10-149-82-88.ec2.internal podname=retina-agent-4hldd version=v0.0.1 apiserver=https://172.20.0.1:443 plugins=packetforward error="failed to reconcile plugin packetforward: fork/exec /bin/clang: no such file or directory" errorVerbose="fork/exec /bin/clang: no such file or directory\nfailed to reconcile plugin packetforward\ngithub.com/microsoft/retina/pkg/managers/pluginmanager.(*PluginManager).Start\n\t/go/src/github.com/microsoft/retina/pkg/managers/pluginmanager/pluginmanager.go:169\ngithub.com/microsoft/retina/pkg/managers/controllermanager.(*Controller).Start.func1\n\t/go/src/github.com/microsoft/retina/pkg/managers/controllermanager/controllermanager.go:108\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\t/go/pkg/mod/golang.org/x/sync@v0.6.0/errgroup/errgroup.go:78\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_arm64.s:1197"
panic: Error running controller manager [recovered]
	panic: Error running controller manager

goroutine 102 [running]:
github.com/microsoft/retina/pkg/telemetry.TrackPanic()
	/go/src/github.com/microsoft/retina/pkg/telemetry/telemetry.go:112 +0x1e8
panic({0x1eb6b00?, 0x4000334180?})
	/usr/local/go/src/runtime/panic.go:914 +0x218
go.uber.org/zap/zapcore.CheckWriteAction.OnWrite(0x0?, 0x1?, {0x4000aaace8?, 0x0?, 0x0?})
	/go/pkg/mod/go.uber.org/zap@v1.26.0/zapcore/entry.go:196 +0x78
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0x40002341a0, {0x400007ff00, 0x1, 0x1})
	/go/pkg/mod/go.uber.org/zap@v1.26.0/zapcore/entry.go:262 +0x2c0
go.uber.org/zap.(*Logger).Panic(0x40005f2080?, {0x25cf876?, 0x0?}, {0x400007ff00, 0x1, 0x1})
	/go/pkg/mod/go.uber.org/zap@v1.26.0/logger.go:284 +0x54
github.com/microsoft/retina/pkg/managers/controllermanager.(*Controller).Start(0x4000cbf450, {0x298b488?, 0x400046f450?})
	/go/src/github.com/microsoft/retina/pkg/managers/controllermanager/controllermanager.go:118 +0x22c
created by main.main in goroutine 1
	/go/src/github.com/microsoft/retina/controller/main.go:286 +0x2190
```

- OS: AmzonLinux
- Kubernetes Version: 1.28
- Host: EKS
- Retina Version: 1.0.0

I am not running any operator. for now I am running only DaemonSet and in configmap I am adding plugins 

```
config.yaml: |-
    apiServer:
      host: 0.0.0.0
      port: 10093
    logLevel: debug
    enabledPlugin: ["packetforward"]
    metricsInterval: 10
    enableTelemetry: false
    enablePodLevel: false
    remoteContext: false
    enableAnnotations: false
```

### PR DESCRIPTION
**Pull request by rbtr**
# Description

This should address the arm/amd64 binary errors in the built image while keeping the Go build stage fast by using the BUILDPLATFORM/TARGETPLATFORM and cross-compiling.

It also moves the tools stage from bullseye to bookworm: bookworm has clang-14 available in the package manager and directly installable, so all of the manual downloads are removed, and the explicit installations are cut down significantly.

Removes some unnecessary Docker cruft which may have been well-intentioned but isn't useful.

## Related Issue

This builds on #260 which fixed #130.

## Checklist

- [x] I have read the [contributing documentation](https://retina.sh/docs/contributing).
- [x] I signed and signed-off the commits (`git commit -S -s ...`). See [this documentation](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification) on signing commits.
- [x] I have correctly attributed the author(s) of the code.
- [x] I have tested the changes locally.
- [x] I have followed the project's style guidelines.
- [x] I have updated the documentation, if necessary.
- [x] I have added tests, if applicable.

## Screenshots (if applicable) or Testing Completed

Please add any relevant screenshots or GIFs to showcase the changes made.

## Additional Notes

Add any additional notes or context about the pull request here.

---

Please refer to the [CONTRIBUTING.md](../CONTRIBUTING.md) file for more information on how to contribute to this project.

### CODE REVIEW

File: controller/Dockerfile.controller [hunk 1]
```diff
@@ -1,136 +1,56 @@
-ARG builderImage="ghcr.io/microsoft/retina/retina-builder:0.0.1"
-ARG toolsImage="ghcr.io/microsoft/retina/retina-tools:0.0.1"
+# go binaries
+FROM --platform=$BUILDPLATFORM mcr.microsoft.com/oss/go/microsoft/golang:1.21 AS gobins
```
Comment (nddq) [inline controller/Dockerfile.controller:2]: do you want to bump this to 1.22 now or wait for [this ](https://github.com/microsoft/retina/pull/301)to go in?

Comment (rbtr) [reply controller/Dockerfile.controller:2]: can bump it now, I am about to push one more change here anyway which collapses the win/linux dockerfiles together

**Comment by rbtr** [comment]
tested by deploying the image from my branch on to an amd/arm+windows cluster and verifying that all retina-agents started and plugins loaded without error
cc @anubhabMajumdar

**Comment by rbtr** [comment]
> what was the cause of the segfault? @rbtr

upgrade the tools image to debian 12 (and installed clang14 there) but the agent image was still using debian 11 distroless. updated that and it resolved

**Review by nddq — APPROVED** [review]

**Pull request merged**